### PR TITLE
Add Rusanov Flux

### PR DESCRIPTION
**Pull request by davschneller**
An ad-hoc implementation based on @sebwolf-de 's centered flux implementation. I.e. you can now choose between using Godunov and Rusanov; both just using only near the fault (faces with adjacent cells having at least one DR face), or everywhere.

### CODE REVIEW

File: Documentation/parameters.par [hunk 1]
```diff
@@ -16,6 +16,9 @@ ITMStartingtime = 2.0 ! Time when ITM is turned on
 ITMTime = 0.01 !Time duration for which ITM is on
 ITMVelocityscalingfactor = 2 ! Scaling factor
 ITMReflectionType = 1 ! 1 reflects both waves, 2 reflects both waves with constant velocity, 3 reflects only P-waves, 4 reflects only S-waves
+
+numflux = 'godunov' ! The numerical flux near the fault. Values: 'godunov', 'rusanov'
```
Comment (Thomas-Ulrich) [inline Documentation/parameters.par:20]: remove "near the fault"
(and the line below precise
The flux for cells adjacent to the fault, excluding the fault boundary flux.
or something like that?)

File: generated_code/aderdg.py [hunk 2]
```diff
@@ -197,11 +197,15 @@ def starMatrix(self, dim):
     pass
 
   def addInit(self, generator):
+    godunov_spp = self.godunov_spp()
+    self.QcorrLocal = Tensor('QcorrLocal', godunov_spp.shape)
+    self.QcorrNeighbor = Tensor('QcorrNeighbor', godunov_spp.shape)
+
     fluxScale = Scalar('fluxScale')
-    computeFluxSolverLocal = self.AplusT['ij'] <= fluxScale * self.Tinv['ki'] * self.QgodLocal['kq'] * self.starMatrix(0)['ql'] * self.T['jl']
+    computeFluxSolverLocal = self.AplusT['ij'] <= fluxScale * self.Tinv['ki'] * (self.QgodLocal['kq'] * self.starMatrix(0)['ql'] + self.QcorrLocal['kl']) * self.T['jl']
```
Comment (Thomas-Ulrich) [inline generated_code/aderdg.py:205]: how do adding these zeros in the case of Godunov ( extra flops) will affect performance?

File: src/Initializer/CellLocalMatrices.cpp [hunk 3]
```diff
@@ -196,25 +203,73 @@ void seissol::initializer::initializeCellLocalMatrices( seissol::geometry::MeshR
...
+        const auto wavespeedNeighbor = material[cell].neighbor[side].getMaxWaveSpeed();
+        const auto wavespeed = std::max(wavespeedLocal, wavespeedNeighbor);
+
+        real centralFluxData[tensor::QgodLocal::size()]{};
+        real rusanovPlusData[tensor::QcorrLocal::size()]{};
+        real rusanovMinusData[tensor::QcorrNeighbor::size()]{};
+        auto centralFluxView = init::QgodLocal::view::create(centralFluxData);
+        auto rusanovPlusView = init::QcorrLocal::view::create(rusanovPlusData);
+        auto rusanovMinusView = init::QcorrNeighbor::view::create(rusanovMinusData);
+        for (size_t i = 0; i < 9; i++) {
```
Comment (Thomas-Ulrich) [inline src/Initializer/CellLocalMatrices.cpp:233]: what about attenuation or poroelasticity?
(should you raise an error maybe?)

**Review by Thomas-Ulrich — APPROVED** [review]
LGTM, but would be nice with Zihua could have a look

**Review by Thomas-Ulrich — COMMENTED** [review]
actually changed "Approve" to "Comment" because Rusanov does not give results that look similar as Godunov when running tpv5 (e.g. missing free surface reflection), while it works with Zihua's branch. so there still might be something fishy, right?
Ha wait, do this flux overprints the free-surface by chance? Probably not, but maybe those cells with at least one free-surface boundary conditions should not be modified?

File: generated_code/aderdg.py (hunk 2 above)
Comment (davschneller) [reply generated_code/aderdg.py:205]: That's only done once during the initialization procedure. The actual computation is not modified; we effectively "just" exchange some matrices.

File: src/Initializer/CellLocalMatrices.cpp (hunk 3 above)
Comment (davschneller) [reply src/Initializer/CellLocalMatrices.cpp:233]: That's why it's more or less still a draft right now. :)

(didn't get around to modifying for those systems yet)

File: src/Initializer/CellLocalMatrices.cpp [hunk 4]
```diff
@@ -196,25 +203,78 @@ void seissol::initializer::initializeCellLocalMatrices( seissol::geometry::MeshR
...
+          rusanovPlusView(i, i) = wavespeed * 0.5;
+          rusanovMinusView(i, i) = -wavespeed * 0.5;
+        }
+
+        const auto fluxDefault = isSpecialBC(side) ?
+          modelParameters.fluxNearFault : modelParameters.flux;
+        
+        const auto flux = 
+          cellInformation[cell].faceTypes[side] == FaceType::FreeSurface ||
+          cellInformation[cell].faceTypes[side] == FaceType::FreeSurfaceGravity
```
Comment (Thomas-Ulrich) [inline src/Initializer/CellLocalMatrices.cpp:244]: probably also absorbing?

Comment (davschneller) [reply src/Initializer/CellLocalMatrices.cpp]: Done; added

**Comment by Aangniu** [comment]
https://github.com/SeisSol/SeisSol/pull/1176/files#diff-554fea388baa28e948a2a1ddb8875411390684c325e972eeceb6586861590357R226

**Review by Aangniu — APPROVED** [review]

**Comment by davschneller** [comment]
> https://github.com/SeisSol/SeisSol/pull/1176/files#diff-554fea388baa28e948a2a1ddb8875411390684c325e972eeceb6586861590357R226
> 
> This coefficient for Rusanov flux, whether to use max between cells. I suggest to use average.
> 
> I have a reference here: https://www.sam.math.ethz.ch/sam_reports/reports_final/reports2016/2016-57.pdf

By the formula (given for staggered grids there; but I'd guess it could translate back? My knowledge with those is a bit lacking still), we'd need to take the average of the material parameters then, and subsequently compute the maximum wavespeed of that? Or maybe the eigenvalue average as approximation?

Can both be implemented as well (but it might take a bit longer than today alas :) ).

**Pull request merged**